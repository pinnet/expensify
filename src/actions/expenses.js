import uuid from 'uuid';
export const addExpense = ({description = '',note = '',amount = 0,createdAt= new Date().getTime() }={}) => ({
    type: 'ADD_EXPENSE',
    expenses:{
        id: uuid(),
        description,
        amount,
        note,
        createdAt
    }  
});

export const removeExpense = ({id} = {}) => ({   
    type: 'REMOVE_EXPENSE',
    id
});

export const editExpense = (id,updates) => ({   
    type: 'EDIT_EXPENSE',
    id,
    updates
});

