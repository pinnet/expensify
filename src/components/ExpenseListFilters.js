import React from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import {DateRangePicker} from 'react-dates';
import {
    setTextFilter,
    sortByAmount,
    sortByDate,
    setStartDate,
    setEndDate} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        callendarFocused:null
    };
    onFocusChange = (callendarFocused) => {
        this.setState({callendarFocused});
    };
    onDatesChange = ({startDate,endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }
    onOptionChange = (e) => {
        if (e.target.value === 'date'){
            this.props.sortByDate();  
        }
        else if (e.target.value === 'amount'){
            this.props.sortByAmount();
        }  
    }
    render(){  
        return (  
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            className="text-input"
                            placeholder="Search expenses"
                            type="text" 
                            value={this.props.filters.text} 
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select 
                            className="select"
                            value = {this.props.filters.sortBy}        
                            onChange={this.onOptionChange} 
                        >
                            <option value="date">Date    </option>
                            <option value="amount">Amount  </option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.callendarFocused}
                            onFocusChange={this.onFocusChange}
                            displayFormat="Do MMM YYYY"
                            numberOfMonths={1}
                            isOutsideRange={(day) => false}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>
        );
    };
}
const mapStateToProps = (state) => ({ filters: state.filters });
const mapDispatchToProps = (dispatch) => ({

    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter:(text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())

});
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);