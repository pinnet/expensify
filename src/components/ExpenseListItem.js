import React from 'react';
import moment from 'moment';
//import {connect} from 'react-redux';
import {Link} from 'react-router-dom';


const ExpenseListItem = ({id,description,amount,createdAt}) => (
 <div>
     <Link to={`/edit/${id}`}><h3>{description} : amount Â£{(amount / 100).toFixed(2)} : created {moment(createdAt).format('Do MMM YYYY')}</h3></Link>
 </div>
);
export default ExpenseListItem;