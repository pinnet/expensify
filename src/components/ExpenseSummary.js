import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses.js';
import getExpensesTotal from '../selectors/expensese-total';

export const ExpensesSummary = (props) => {
    if (props.expenses !== undefined){
        const expenseCount = props.expenses.length;
        const expensesTotal = getExpensesTotal(props.expenses);
        return(
            <div className="page-header"> 
                <div className="content-container">
                    {props.expenses.length === 0 ||
                    <h1 className="page-header__title">Viewing <span>{expenseCount},</span> expense{expenseCount === 1 || 's'} totaling <span>Â£{numeral(expensesTotal / 100).format('0,0.00')}</span>.</h1>
                    }
                    <div className="page_header__actions">
                        <Link className="button" to="/create">Add Expense</Link>
                    </div>
                </div>
            </div>
        )
    }
};

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses,state.filters)
});
export default connect(mapStateToProps)(ExpensesSummary);