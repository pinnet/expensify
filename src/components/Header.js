import React from 'react';
import {NavLink} from 'react-router-dom';


const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink> |&nbsp; 
        <NavLink to="/create" activeClassName="is-active">Create New Expense</NavLink> |&nbsp; 
        <NavLink to="/logout" activeClassName="is-active">logout</NavLink>
    </header>
);
export default Header;