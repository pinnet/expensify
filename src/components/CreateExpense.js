import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class CreateExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        if(this.props.history){
            this.props.history.push('/');
        }
    };
    render() {
        return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1>Add Expense</h1> 
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        </div>    
        );
    };
}
const mapDispatchToProps = (dispatch) =>({   
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});
export default connect(undefined,mapDispatchToProps)(CreateExpense)