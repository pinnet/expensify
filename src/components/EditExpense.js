import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id,expense);
        if(this.props.history){
            this.props.history.push('/');
        }
    };
    onClick = () => {
        this.props.startRemoveExpense({id:this.props.expense.id});
        if(this.props.history){
            this.props.history.push('/');
        }
    };
    render(){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-headder__title">Edit Expense</h1> 
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                      expense={this.props.expense}
                      onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onClick}>Remove Expense</button>
                </div>
                
            </div>
            );
    }
}

const mapStateToProps = (state,props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});
const mapDispatchToProps = (dispatch) => ({
    startEditExpense:(id,expense) => dispatch(startEditExpense(id,expense)),
    startRemoveExpense:(data) => dispatch(startRemoveExpense(data))
});
   
export default connect(mapStateToProps,mapDispatchToProps)(EditExpense);