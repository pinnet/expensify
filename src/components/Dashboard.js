import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpenseSummary';
import CreateExpense from './CreateExpense';

const Dashboard = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList/>   
    </div>
)
export default Dashboard;