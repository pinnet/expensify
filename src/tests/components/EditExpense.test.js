import React from 'react';
import { shallow } from 'enzyme'; 
import moment from 'moment';
import expenses from '../fixtures/expenses';
import {EditExpense} from '../../components/EditExpense';

let editExpense,removeExpense, history, wrapper;
beforeEach(() => {

    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpense
        editExpense={editExpense} 
        removeExpense={removeExpense}
        history={history}
        expense={expenses[0]}/>);
});

it('should render editExpense correctly ',() => {
    expect(wrapper).toMatchSnapshot(); 
});

it('should handle editExpense', () => {
    
        wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
        expect(history.push).toHaveBeenLastCalledWith('/');
        expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,expenses[0]);
    
});
it('should handle removeExpense', () => {
    
        wrapper.find('button').simulate('click');
        expect(history.push).toHaveBeenLastCalledWith('/');
        expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[0].id});
    
});