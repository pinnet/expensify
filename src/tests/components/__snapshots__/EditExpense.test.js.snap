// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render editExpense correctly  1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-headder__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": "1970-01-01T00:00:00.000Z",
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
